File: projectforge-business/src/main/java/org/projectforge/framework/persistence/history/HistoryBaseDaoAdapter.java
package org.projectforge.framework.persistence.history;

import java.io.Serializable;
import java.lang.reflect.AccessibleObject;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import org.apache.commons.lang3.ObjectUtils;
import org.apache.log4j.Logger;
import org.projectforge.business.user.UserGroupCache;
import org.projectforge.framework.configuration.ApplicationContextProvider;
import org.projectforge.framework.persistence.api.BaseDO;
import org.projectforge.framework.persistence.api.ExtendedBaseDO;
import org.projectforge.framework.persistence.api.ModificationStatus;
import org.projectforge.framework.persistence.api.PFPersistancyBehavior;
import org.projectforge.framework.persistence.jpa.PfEmgrFactory;

import de.micromata.genome.db.jpa.history.api.DiffEntry;
import de.micromata.genome.db.jpa.history.api.HistProp;
import de.micromata.genome.db.jpa.history.api.HistoryEntry;
import de.micromata.genome.db.jpa.history.api.HistoryService;
import de.micromata.genome.db.jpa.history.api.HistoryServiceManager;
import de.micromata.genome.db.jpa.history.api.WithHistory;
import de.micromata.genome.db.jpa.history.entities.EntityOpType;
import de.micromata.genome.db.jpa.history.impl.HistoryEmgrAfterInsertedEventHandler;
import de.micromata.genome.db.jpa.history.impl.HistoryUpdateCopyFilterEventListener;
import de.micromata.genome.jpa.DbRecord;
import de.micromata.genome.jpa.events.EmgrAfterInsertedEvent;
import de.micromata.genome.jpa.events.EmgrUpdateCopyFilterEvent;
import de.micromata.genome.util.runtime.ClassUtils;
import de.micromata.hibernate.history.delta.PropertyDelta;
import de.micromata.hibernate.history.delta.SimplePropertyDelta;

/**
 * Utility to provide compat with BaseDao.
 *
 * @author Roger Rene Kommer (r.kommer.extern@micromata.de)
 */
public class HistoryBaseDaoAdapter
{
  private static final Logger log = Logger.getLogger(HistoryBaseDaoAdapter.class);

  private static final HistoryEntry[] HISTORY_ARR_TEMPL = new HistoryEntry[] {};

  public static HistoryEntry[] getHistoryFor(BaseDO<?> obj)
  {
    long begin = System.currentTimeMillis();
    HistoryEntry[] result = getHistoryEntries(obj).toArray(HISTORY_ARR_TEMPL);
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.getHistoryFor took: " + (end - begin) + " ms.");
    return result;
  }

  public static List<? extends HistoryEntry> getHistoryEntries(BaseDO<?> ob)
  {
    long begin = System.currentTimeMillis();
    HistoryService histservice = HistoryServiceManager.get().getHistoryService();
    PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
    List<? extends HistoryEntry> ret = emf.runInTrans((emgr) -> {
      return histservice.getHistoryEntries(emgr, ob);
    });
    List<? extends HistoryEntry> nret = ret.stream()
        .sorted((e1, e2) -> e2.getModifiedAt().compareTo(e1.getModifiedAt())).collect(Collectors.toList());
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.getHistoryEntries took: " + (end - begin) + " ms.");
    return nret;
  }

  public static PropertyDelta diffEntryToPropertyDelta(DiffEntry de)
  {
    long begin = System.currentTimeMillis();
    SimplePropertyDelta ret = new SimplePropertyDelta(de.getPropertyName(), String.class, de.getOldValue(),
        de.getNewValue());
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.diffEntryToPropertyDelta took: " + (end - begin) + " ms.");
    return ret;
  }

  public static List<SimpleHistoryEntry> getSimpleHistoryEntries(final BaseDO<?> ob, UserGroupCache userGroupCache)
  {
    long begin = System.currentTimeMillis();
    List<SimpleHistoryEntry> ret = new ArrayList<>();
    List<? extends HistoryEntry> hel = getHistoryEntries(ob);

    for (HistoryEntry he : hel) {
      List<DiffEntry> deltas = he.getDiffEntries();
      if (deltas.isEmpty() == true) {
        SimpleHistoryEntry se = new SimpleHistoryEntry(userGroupCache, he);
        ret.add(se);
      } else {
        for (DiffEntry de : deltas) {
          final SimpleHistoryEntry se = new SimpleHistoryEntry(userGroupCache, he, diffEntryToPropertyDelta(de));
          ret.add(se);
        }
      }

    }
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.getSimpleHistoryEntries took: " + (end - begin) + " ms.");
    return ret;
  }

  public static boolean isHistorizable(Object bean)
  {
    if (bean == null) {
      return false;
    }
    return isHistorizable(bean.getClass());
  }

  public static boolean isHistorizable(Class<?> clazz)
  {
    long begin = System.currentTimeMillis();
    boolean result = HistoryServiceManager.get().getHistoryService().hasHistory(clazz);
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.isHistorizable took: " + (end - begin) + " ms.");
    return result;
  }

  private static String histCollectionValueToString(Class<?> valueClass, Collection<?> value)
  {
    StringBuilder sb = new StringBuilder();
    for (Object ob : value) {
      if (sb.length() > 0) {
        sb.append(",");
      }
      if (ob instanceof DbRecord) {
        DbRecord rec = (DbRecord) ob;
        sb.append(rec.getPk());
      } else {
        sb.append(ObjectUtils.toString(ob));
      }
    }
    return sb.toString();
  }

  private static String histValueToString(Class<?> valueClass, Object value)
  {
    if (value == null) {
      return null;
    }
    if (value instanceof Collection) {
      return histCollectionValueToString(valueClass, (Collection) value);
    }
    return ObjectUtils.toString(value);
  }

  public static void createHistoryEntry(Object entity, Number id, String user, String property,
      Class<?> valueClass, Object oldValue, Object newValue)
  {
    long begin = System.currentTimeMillis();
    String oldVals = histValueToString(valueClass, oldValue);
    String newVals = histValueToString(valueClass, newValue);

    PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
    emf.runInTrans((emgr) -> {
      HistoryServiceManager.get().getHistoryService().insertManualEntry(emgr, EntityOpType.Update,
          entity.getClass().getName(),
          id, user, property, valueClass.getName(), oldVals, newVals);
      return null;
    });
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.createHistoryEntry took: " + (end - begin) + " ms.");
  }

  public static void inserted(BaseDO<?> ob)
  {
    long begin = System.currentTimeMillis();
    PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
    emf.runInTrans((emgr) -> {
      EmgrAfterInsertedEvent event = new EmgrAfterInsertedEvent(emgr, ob);
      new HistoryEmgrAfterInsertedEventHandler().onEvent(event);
      return null;
    });
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.inserted took: " + (end - begin) + " ms.");
  }

  public static ModificationStatus wrappHistoryUpdate(BaseDO<?> dbo, Supplier<ModificationStatus> callback)
  {
    long begin = System.currentTimeMillis();
    final HistoryService historyService = HistoryServiceManager.get().getHistoryService();
    final List<WithHistory> whanots = historyService.internalFindWithHistoryEntity(dbo);
    if (whanots.isEmpty() == true) {
      return callback.get();
    }

    final List<BaseDO<?>> entitiesToHistoricize = getSubEntitiesToHistoricizeDeep(dbo);
    final PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
    final ModificationStatus result = emf.runInTrans((emgr) -> {
      final Map<Serializable, HistoryProperties> props = new HashMap<>();

      // get the (old) history properties before the modification
      entitiesToHistoricize.forEach(
          entity -> {
            final HistoryProperties p = getOrCreateHistoryProperties(props, entity);
            p.oldProps = historyService.internalGetPropertiesForHistory(emgr, whanots, entity);
          }
      );

      // do the modification
      final ModificationStatus ret = callback.get();

      // get the (new) history properties after the modification
      entitiesToHistoricize.forEach(
          entity -> {
            final HistoryProperties p = getOrCreateHistoryProperties(props, entity);
            p.newProps = historyService.internalGetPropertiesForHistory(emgr, whanots, entity);
          }
      );

      // create history entries with the diff resulting from the old and new history properties
      props.forEach(
          (pk, p) -> {
            if (p.oldProps != null && p.newProps != null) {
              historyService.internalOnUpdate(emgr, p.entClassName, pk, p.oldProps, p.newProps);
            }
          }
      );
      return ret;
    });

    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.wrappHistoryUpdate took: " + (end - begin) + " ms.");
    return result;
  }

  /**
   * Nested class just to hold some temporary history data.
   */
  private static final class HistoryProperties
  {
    private String entClassName;
    private Map<String, HistProp> oldProps;
    private Map<String, HistProp> newProps;
  }

  private static HistoryProperties getOrCreateHistoryProperties(final Map<Serializable, HistoryProperties> props, final DbRecord<?> entity)
  {
    final Serializable pk = entity.getPk();
    if (props.containsKey(pk)) {
      return props.get(pk);
    } else {
      final HistoryProperties hp = new HistoryProperties();
      props.put(pk, hp);
      hp.entClassName = entity.getClass().getName();
      return hp;
    }
  }

  private static List<BaseDO<?>> getSubEntitiesToHistoricizeDeep(final BaseDO<?> entity)
  {
    final List<BaseDO<?>> result = new ArrayList<>();
    final Queue<BaseDO<?>> queue = new LinkedList<>();
    queue.add(entity);

    // do a breadth first search through the tree
    while (!queue.isEmpty()) {
      final BaseDO<?> head = queue.poll();
      result.add(head);
      final List<BaseDO<?>> subEntries = getSubEntitiesToHistoricize(head);
      queue.addAll(subEntries);
    }

    return result;
  }

  /**
   * Takes a DO and returns a list of DOs. This list contains all entries of the collections of the DOs where the class fields have this annotation:
   * "@PFPersistancyBehavior(autoUpdateCollectionEntries = true)".
   *
   * @param entity The DO.
   * @return The List of DOs.
   */
  private static List<BaseDO<?>> getSubEntitiesToHistoricize(final BaseDO<?> entity)
  {
    final Collection<Field> fields = ClassUtils.getAllFields(entity.getClass()).values();
    AccessibleObject.setAccessible(fields.toArray(new Field[0]), true);

    return fields
        .stream()
        .filter(field -> {
          final PFPersistancyBehavior behavior = field.getAnnotation(PFPersistancyBehavior.class);
          return behavior != null && behavior.autoUpdateCollectionEntries();
        })
        .map(field -> {
          try {
            return (Collection<BaseDO<?>>) field.get(entity);
          } catch (IllegalAccessException | ClassCastException e) {
            return (Collection<BaseDO<?>>) Collections.EMPTY_LIST;
          }
        })
        .flatMap(Collection::stream)
        .collect(Collectors.toList());
  }

  public static void updated(BaseDO<?> oldo, BaseDO<?> newo)
  {
    long begin = System.currentTimeMillis();
    PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
    emf.runInTrans((emgr) -> {
      EmgrUpdateCopyFilterEvent event = new EmgrUpdateCopyFilterEvent(emgr, oldo.getClass(), oldo.getClass(), oldo,
          newo,
          true);
      new HistoryUpdateCopyFilterEventListener().onEvent(event);
      return null;
    });
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.updated took: " + (end - begin) + " ms.");
  }

  public static void markedAsDeleted(ExtendedBaseDO<?> oldo, ExtendedBaseDO<?> newoj)
  {
    long begin = System.currentTimeMillis();
    boolean prev = newoj.isDeleted();
    newoj.setDeleted(true);
    updated(oldo, newoj);
    newoj.setDeleted(prev);
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.markedAsDeleted took: " + (end - begin) + " ms.");
  }

  public static void markedAsUnDeleted(ExtendedBaseDO<?> oldo, ExtendedBaseDO<?> newoj)
  {
    long begin = System.currentTimeMillis();
    boolean prev = newoj.isDeleted();
    newoj.setDeleted(false);
    updated(oldo, newoj);
    newoj.setDeleted(prev);
    long end = System.currentTimeMillis();
    log.info("HistoryBaseDaoAdapter.markedAsUnDeleted took: " + (end - begin) + " ms.");
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/framework/persistence/history/HistoryBaseDaoAdapter.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2025 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.framework.persistence.history

import org.projectforge.common.AnnotationsUtils
import org.projectforge.framework.persistence.api.BaseDO
import org.projectforge.framework.persistence.api.IdObject
import org.projectforge.framework.persistence.candh.CandHHistoryEntryICustomizer
import org.projectforge.framework.persistence.entities.AbstractHistorizableBaseDO
import org.projectforge.framework.persistence.jpa.PfPersistenceContext
import java.util.*

/**
 * Utility to provide functionalities for BaseDao.
 *
 * @author Roger Rene Kommer (r.kommer.extern@micromata.de)
 * @author Kai Reinhard
 */
object HistoryBaseDaoAdapter {
    fun isHistorizable(bean: Any?): Boolean {
        if (bean == null) {
            return false
        }
        return isHistorizable(bean.javaClass)
    }

    fun isHistorizable(entityClass: Class<*>): Boolean {
        if (AbstractHistorizableBaseDO::class.java.isAssignableFrom(entityClass)) {
            return true
        }
        return AnnotationsUtils.hasClassAnnotation(entityClass, WithHistory::class.java)
    }

    fun createHistoryEntry(entity: IdObject<Long>, opType: EntityOpType): HistoryEntryDO {
        return HistoryEntryDO.create(entity, opType)
    }

    /**
     * Convention: If you want to create a history entry of collections, the oldValue should contain all elements that are removed and the newValue should contain all elements that are added.
     * @param oldValue Supports all types supported by [HistoryValueHandlerRegistry]. Also, collections of objects are supported and will be serialized to a csv string.
     * @param newValue Supports all types supported by [HistoryValueHandlerRegistry]. Also, collections of objects are supported and will be serialized to a csv string.
     */
    fun createHistoryUpdateEntryWithSingleAttribute(
        entity: IdObject<Long>,
        propertyName: String?,
        propertyTypeClass: Class<*>,
        oldValue: Any?,
        newValue: Any?,
    ): HistoryEntryDO {
        val entry = HistoryEntryDO.create(entity, EntityOpType.Update)
        HistoryEntryAttrDO.create(
            propertyTypeClass = propertyTypeClass,
            opType = PropertyOpType.Update,
            oldValue = oldValue,
            newValue = newValue,
            propertyName = propertyName,
            historyEntry = entry,
        )

        return entry
    }

    private fun insertHistoryEntry(
        entity: Any,
        historyEntry: HistoryEntryDO,
        context: PfPersistenceContext,
    ): HistoryEntryDO {
        if (entity is CandHHistoryEntryICustomizer) {
            entity.customize(historyEntry)
        }
        context.insert(historyEntry)
        return historyEntry
    }

    /**
     * Convention: If you want to create a history entry of collections, the oldValue should contain all elements that are removed and the newValue should contain all elements that are added.
     * @param oldValue Supports all types supported by [HistoryValueHandlerRegistry]. Also, collections of objects are supported and will be serialized to a csv string.
     * @param newValue Supports all types supported by [HistoryValueHandlerRegistry]. Also, collections of objects are supported and will be serialized to a csv string.
     */
    fun insertHistoryUpdateEntryWithSingleAttribute(
        entity: IdObject<Long>,
        propertyName: String?,
        propertyTypeClass: Class<*>,
        oldValue: Any?, newValue: Any?,
        context: PfPersistenceContext,
    ): HistoryEntryDO {
        val historyEntry = createHistoryUpdateEntryWithSingleAttribute(
            entity, propertyName, propertyTypeClass, oldValue, newValue,
        )
        return insertHistoryEntry(entity, historyEntry, context)
    }

    fun inserted(obj: BaseDO<Long>, context: PfPersistenceContext) {
        if (!isHistorizable(obj)) {
            // not historizable
            return
        }
        val historyEntry = createHistoryEntry(obj, EntityOpType.Insert)
        insertHistoryEntry(obj, historyEntry, context)
    }

    fun updated(
        obj: BaseDO<Long>,
        historyEntries: List<HistoryEntryDO>?,
        context: PfPersistenceContext
    ) {
        if (!isHistorizable(obj) || historyEntries.isNullOrEmpty()) {
            return
        }
        historyEntries.forEach { entry ->
            insertHistoryEntry(obj, entry, context)
            /*entry.attributes?.forEach { attr ->
                entry.add(attr)
                attr.internalSerializeValueObjects()
                context.insert(attr)
            }*/
        }
    }

    /*
    public static ModificationStatus wrapHistoryUpdate(BaseDO<?> dbo, Supplier<ModificationStatus> callback) {
        final HistoryService historyService = HistoryServiceManager.get().getHistoryService();
        final List<WithHistory> whanots = historyService.internalFindWithHistoryEntity(dbo);
        if (whanots.isEmpty()) {
            return callback.get();
        }
        final List<BaseDO<?>> entitiesToHistoricize = getSubEntitiesToHistoricizeDeep(dbo);
        final PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
        final ModificationStatus result = emf.runInTrans((emgr) -> {
            return wrapHistoryUpdate(emgr, historyService, whanots, dbo, callback);
        });

        //long end = System.currentTimeMillis();
        //log.info("HistoryBaseDaoAdapter.wrappHistoryUpdate took: " + (end - begin) + " ms.");
        return result;
    }

    public static ModificationStatus wrapHistoryUpdate(PfEmgr emgr, BaseDO<?> dbo, Supplier<ModificationStatus> callback) {
        final HistoryService historyService = HistoryServiceManager.get().getHistoryService();
        final List<WithHistory> whanots = historyService.internalFindWithHistoryEntity(dbo);
        if (whanots.isEmpty()) {
            return callback.get();
        }
        return wrapHistoryUpdate(emgr, historyService, whanots, dbo, callback);
    }

    private static ModificationStatus wrapHistoryUpdate(PfEmgr emgr, HistoryService historyService, List<WithHistory> whanots, BaseDO<?> dbo, Supplier<ModificationStatus> callback) {
        final List<BaseDO<?>> entitiesToHistoricize = getSubEntitiesToHistoricizeDeep(dbo);
        final Map<Serializable, HistoryProperties> props = new HashMap<>();

        // get the (old) history properties before the modification
        entitiesToHistoricize.forEach(
                entity -> {
                    final HistoryProperties p = getOrCreateHistoryProperties(props, entity);
                    p.oldProps = historyService.internalGetPropertiesForHistory(emgr, whanots, entity);
                }
        );

        // do the modification
        final ModificationStatus result = callback.get();

        // get the (new) history properties after the modification
        entitiesToHistoricize.forEach(
                entity -> {
                    final HistoryProperties p = getOrCreateHistoryProperties(props, entity);
                    p.newProps = historyService.internalGetPropertiesForHistory(emgr, whanots, entity);
                }
        );

        // create history entries with the diff resulting from the old and new history properties
        props.forEach(
                (pk, p) -> {
                    if (p.oldProps != null && p.newProps != null) {
                        try {
                            historyService.internalOnUpdate(emgr, p.entClassName, pk, p.oldProps, p.newProps);
                        } catch (Exception ex) {
                            log.error("Error while writing history entry (" + p.entClassName + ":" + pk + ", '" + p.oldProps + "'->'" + p.newProps + "': " + ex.getMessage(), ex);
                        }
                    }
                }
        );

        //long end = System.currentTimeMillis();
        //log.info("HistoryBaseDaoAdapter.wrappHistoryUpdate took: " + (end - begin) + " ms.");
        return result;
    }*/
    /**
     * Nested class just to hold some temporary history data.
     * /
     * private static final class HistoryProperties {
     * private String entClassName;
     * private Map<String></String>, HistProp> oldProps;
     * private Map<String></String>, HistProp> newProps;
     * }
     *
     * private static HistoryProperties getOrCreateHistoryProperties(final Map<Serializable></Serializable>, HistoryProperties> props, final DbRecord entity) {
     * final Serializable pk = entity.getPk();
     * if (props.containsKey(pk)) {
     * return props.get(pk);
     * } else {
     * final HistoryProperties hp = new HistoryProperties();
     * props.put(pk, hp);
     * hp.entClassName = entity.getClass().getName();
     * return hp;
     * }
     * }
     *
     * private static List<BaseDO></BaseDO>> getSubEntitiesToHistoricizeDeep(final BaseDO entity) {
     * final List<BaseDO></BaseDO>> result = new ArrayList<>();
     * final Queue<BaseDO></BaseDO>> queue = new LinkedList<>();
     * queue.add(entity);
     *
     * // do a breadth first search through the tree
     * while (!queue.isEmpty()) {
     * final BaseDO head = queue.poll();
     * result.add(head);
     * final List<BaseDO></BaseDO>> subEntries = getSubEntitiesToHistoricize(head);
     * queue.addAll(subEntries);
     * }
     *
     * return result;
     * }
     *
     * / **
     * Takes a DO and returns a list of DOs. This list contains all entries of the collections of the DOs where the class fields have this annotation:
     * "@PFPersistancyBehavior(autoUpdateCollectionEntries = true)".
     *
     * @param entity The DO.
     * @return The List of DOs.
     * /
     * private static List<BaseDO></BaseDO>> getSubEntitiesToHistoricize(final BaseDO entity) {
     * final Collection<Field> fields = ClassUtils.getAllFields(entity.getClass()).values();
     * AccessibleObject.setAccessible(fields.toArray(new Field[0]), true);
     *
     * return fields
     * .stream()
     * .filter(field -> {
     * final PFPersistancyBehavior behavior = field.getAnnotation(PFPersistancyBehavior.class);
     * return behavior != null && behavior.autoUpdateCollectionEntries();
     * })
     * .map(field -> {
     * try {
     * return (Collection<BaseDO></BaseDO>>) field.get(entity);
     * } catch (IllegalAccessException | ClassCastException e) {
     * return (Collection<BaseDO></BaseDO>>) Collections.EMPTY_LIST;
     * }
     * })
     * .flatMap(Collection::stream)
     * .collect(Collectors.toList());
     * }
     *
     * public static void updated(BaseDO oldo, BaseDO newo) {
     * //long begin = System.currentTimeMillis();
     * PfEmgrFactory emf = ApplicationContextProvider.getApplicationContext().getBean(PfEmgrFactory.class);
     * emf.runInTrans((emgr) -> {
     * EmgrUpdateCopyFilterEvent event = new EmgrUpdateCopyFilterEvent(emgr, oldo.getClass(), oldo.getClass(), oldo,
     * newo,
     * true);
     * new HistoryUpdateCopyFilterEventListener().onEvent(event);
     * return null;
     * });
     * //long end = System.currentTimeMillis();
     * //log.info("HistoryBaseDaoAdapter.updated took: " + (end - begin) + " ms.");
     * }
     *
     * public static void markedAsDeleted(ExtendedBaseDO oldo, ExtendedBaseDO newoj) {
     * //long begin = System.currentTimeMillis();
     * boolean prev = newoj.isDeleted;
     * newoj.isDeleted = true;
     * updated(oldo, newoj);
     * newoj.isDeleted = prev;
     * //long end = System.currentTimeMillis();
     * //log.info("HistoryBaseDaoAdapter.markedAsDeleted took: " + (end - begin) + " ms.");
     * }
     *
     * public static void markedAsUnDeleted(ExtendedBaseDO oldo, ExtendedBaseDO newoj) {
     * //long begin = System.currentTimeMillis();
     * boolean prev = newoj.isDeleted;
     * newoj.isDeleted = false;
     * updated(oldo, newoj);
     * newoj.isDeleted = prev;
     * //long end = System.currentTimeMillis();
     * //log.info("HistoryBaseDaoAdapter.markedAsUnDeleted took: " + (end - begin) + " ms.");
     * }
    </Field> */
}
